pydicom/dataset.py

```diff
@@ -355,78 +362,164 @@ def __iter__(self):
+                gdcm.PixelFormat.FLOAT64:  numpy.float64
+            }
+            gdcm_pixel_format = gdcm_image.GetPixelFormat().GetScalarType()
+            if gdcm_pixel_format in gdcm_numpy_typemap:
+                numpy_dtype = gdcm_numpy_typemap[gdcm_pixel_format]
+            else:
+                raise TypeError('{} is not a GDCM supported pixel format'.format(gdcm_pixel_format))
+            
+            # get the raw data buffer (decompressed, decoded)
+            gdcm_raw_data_buffer = None
+            
+            # in the case of a lookup table try to apply it
+            if gdcm_image.GetPhotometricInterpretation().GetType() == gdcm.PhotometricInterpretation.PALETTE_COLOR:
+                gdcm_buf_len = gdcm_image.GetBufferLength()
+                # apply lookup table if possible
+                gdcm_lutfilt = gdcm.ImageApplyLookupTable();
+                gdcm_lutfilt.SetInput(gdcm_image);
+                if not gdcm_lutfilt.Apply():
+                    warnings.warn("GDCM failed to apply LUT")
+                gdcm_lutfiltered_pixmap = gdcm_lutfilt.GetOutputAsPixmap()
+                gdcm_raw_data_buffer = gdcm_lutfiltered_pixmap.GetBuffer()
+                
+                # amount of samples per pixel can change
+                self.SamplesPerPixel = gdcm_lutfiltered_pixmap.GetBufferLength() / gdcm_buf_len
+                self.PlanarConfiguration = 0
+            
+            # if we haven't gotten the buffer yet, just get it directly
+            if gdcm_raw_data_buffer is None:
+                gdcm_raw_data_buffer = gdcm_image.GetBuffer()
+            
+            # if GDCM indicates that a byte swap is in order, make sure to inform numpy as well
+            if gdcm_image.GetNeedByteSwap():
+                numpy_dtype.newbyteorder('S')
+
+            # GDCM returns char* as type str. In the case of Python 3 this needs to be encoded back into a byte array.
+            # This is no problem in Python 3 as strings are byte strings by default there.
+            if sys.version_info >= (3, 0):
+                def encode_gdcm_buffer_as_bytearray(gdcm_buffer, encoding=None):
+                    return gdcm_buffer.encode(encoding if encoding is not None else sys.getfilesystemencoding(), "surrogateescape")
+                try:
+                    gdcm_raw_data_buffer = encode_gdcm_buffer_as_bytearray(gdcm_raw_data_buffer)
+                except UnicodeEncodeError:
+                    try:
+                        gdcm_raw_data_buffer = encode_gdcm_buffer_as_bytearray(gdcm_raw_data_buffer, encoding='utf-8')
+                    except UnicodeEncodeError:
+                        try:
+                            gdcm_raw_data_buffer = encode_gdcm_buffer_as_bytearray(gdcm_raw_data_buffer, encoding='mbcs')
+                        except UnicodeEncodeError:
+                            raise TypeError("Could not encode the image unicode string returned by GDCM back to a byte array")
```

Comment:
Short answer: it's complicated.

Long answer: GDCM returns the byte buffer as `char*` in c++. This is returned as a `str` by the Python wrappers generated by SWIG. 

On Python 2 `str` is a byte buffer by default and everything is OK. On Python 3, `str` is a unicode string and no longer a byte buffer. Since we want to access the bytes directly we need to re-encode the unicode string to their original corresponding bytes (a `bytearray` in py3).

This requires us to know the original encoding of the file, which we can only guess, as far as I am aware. This is why we need to try a bunch of codecs. Ideally we would be able to slecet the appropriate codec directly, but I don't know how to do that.

We also need to specify the special surrogate error handler since some bytes have no corresponding unicode character and this would result in an exception under normal conditions.

Also see the SWIG docs on this topic:
http://www.swig.org/Doc3.0/Python.html#Python_nn77

This article also explains a lot about unicode handling in py3:
http://python-notes.curiousefficiency.org/en/latest/python3/text_file_processing.html